python/pyiceberg/io/pyarrow.py

```diff
@@ -532,131 +535,116 @@ def project_table(
 
 
 def to_requested_schema(requested_schema: Schema, file_schema: Schema, table: pa.Table) -> pa.Table:
-    return VisitWithArrow(requested_schema, file_schema, table).visit()
+    struct_array = visit_with_partner(requested_schema, table, ArrowProjectionVisitor(file_schema), ArrowAccessor(file_schema))
 
+    arrays = []
+    fields = []
+    for pos, field in enumerate(requested_schema.fields):
+        array = struct_array.field(pos)
+        arrays.append(array)
+        fields.append(pa.field(field.name, array.type, field.optional))
+    return pa.Table.from_arrays(arrays, schema=pa.schema(fields))
 
-class VisitWithArrow:
-    requested_schema: Schema
+
+class ArrowProjectionVisitor(SchemaWithPartnerVisitor[pa.Array, Optional[pa.Array]]):
     file_schema: Schema
-    table: pa.Table
 
-    def __init__(self, requested_schema: Schema, file_schema: Schema, table: pa.Table) -> None:
-        self.requested_schema = requested_schema
+    def __init__(self, file_schema: Schema):
         self.file_schema = file_schema
-        self.table = table
-
-    def visit(self) -> pa.Table:
-        return self.visit_with_arrow(self.requested_schema, self.file_schema)
-
-    @singledispatchmethod
-    def visit_with_arrow(self, requested_schema: Union[Schema, IcebergType], file_schema: Union[Schema, IcebergType]) -> pa.Table:
-        """A generic function for applying a schema visitor to any point within a schema
 
-        The function traverses the schema in post-order fashion
-
-        Args:
-            obj(Schema | IcebergType): An instance of a Schema or an IcebergType
-            visitor (VisitWithArrow[T]): An instance of an implementation of the generic VisitWithArrow base class
+    def cast_if_needed(self, field: NestedField, values: pa.Array) -> pa.Array:
```

Comment:
Rather than traversing both schemas at the same time and looking up columns, this visits the expected schema and the Arrow arrays at the same time. The only lookup needed is this one to possibly promote fields when they are added to the final Arrow tree.

Reply:
I agree that we're doing a lot of things with that single visitor, but since it was a lookup, I was hesitant to introduce yet another visitor.

Reply:
I don't think we should be hesitant to add more visitors. The purpose of the visitor is to standardize how we express transformations and logic for working with tree structures. It helps a lot if you don't need to think about how the visiting works to understand what's going on.